### PR TITLE
varlink: make polkit auth work with released polkit versions

### PR DESCRIPTION
**Pull request by poettering**

### CODE REVIEW

**Comment by bluca** [comment]
I think we should avoid adding support for authenticating based on pid+start time+uid, as it's way too fraught with issues and overall not reliable. I'll ping the polkit maintainer next week to ask for a review, I hope we can get that merged, as it's a very simple change over the rest of the pidfd work that is already merged. I think it's ok to require new versions of stuff for new features.

**Comment by poettering** [comment]
> I think we should avoid adding support for authenticating based on pid+start time+uid

wel, the way i see it, as long as pk still supports it and hasn't deprecated its use it's ok to use. I think with the inclusion of the "uid" it's acually kinda OK.

>, as it's way too fraught with issues and overall not reliable. I'll ping the polkit maintainer next week to ask for a review, I hope we can get that merged, as it's a very simple change over the rest of the pidfd work that is already merged. I think it's ok to require new versions of stuff for new features.

We sould certainly start using pidfds with this once it gets merged, but the reason why I think supporting the existing triplet is nice, is because it allows us to switch some things we so fr used dbus for to varlink without having to risk compat break with old systems

**Comment by bluca** [comment]
> > I think we should avoid adding support for authenticating based on pid+start time+uid
> 
> wel, the way i see it, as long as pk still supports it and hasn't deprecated its use it's ok to use. I think with the inclusion of the "uid" it's acually kinda OK.
> 
> > , as it's way too fraught with issues and overall not reliable. I'll ping the polkit maintainer next week to ask for a review, I hope we can get that merged, as it's a very simple change over the rest of the pidfd work that is already merged. I think it's ok to require new versions of stuff for new features.
> 
> We sould certainly start using pidfds with this once it gets merged, but the reason why I think supporting the existing triplet is nice, is because it allows us to switch some things we so fr used dbus for to varlink without having to risk compat break with old systems

The polkit PR is now merged, so it will be part of the next release, which is hopefully going to happen soon.

Regarding using process start time to identify a caller, this is what's documented in the current version of polkit:

https://gitlab.freedesktop.org/polkit/polkit/-/blob/123/src/polkit/polkitunixprocess.c?ref_type=tags#L54

```
 * An object for representing a UNIX process.  NOTE: This object as
 * designed is now known broken; a mechanism to exploit a delay in
 * start time in the Linux kernel was identified.  Avoid
 * calling polkit_subject_equal() to compare two processes.
```

There are many more paragraphs of scary warning, borne out of CVEs, following that.

I am not convinced the fallback should be merged. It is very much broken and there are known exploits as polkit has seen. I imagine this functionality will be used to authorize, among other things, loading block devices without privileges via systemd-backdoord - we really, really do not want to ship a broken authentication method for that, as it's pretty much a kernel-breaking escalation path. The pid fd method is safe and proven, and we can use it to ship this feature with confidence that it will work as expected (and if it doesn't, we can just blame the kernel VFS maintainer, easy)

Yes, it will suck that if running with new systemd and old polkit you won't be able to use new features, but such is life: new code with new features requires new dependencies. We don't break backward compatibility, but we do require new versions of stuff for new features, that would not be an unexpected turn of events.

**Comment by poettering** [comment]
The starttime thing is a red herring. Here we authenticate by UID, and that's always safe (and frankly the only thing PK should ever have done...). 

It's a triplet of PID, starttime, UID now, and the former two might be interesting to route the polkit dialog properly, but the last one is where the security is, and what makes sense to authenticate by. Hence I dont really see an issue here.

**Comment by poettering** [comment]
Also, if you really think that there was a security issue, then please get polkit to deprecate auth by uid, it's not really on us to make use of that.

I mean, deprecating authentication by UID sounds more than weird to me, but this is what your proposal boils down to.

**Comment by poettering** [comment]
I think auth-by-proc is pretty useless to me, and is probably the part that should be deprecated if you ask me. auth-by-uid otoh is the part that we should definitely keep, it's the part that just makes sense. And this uses that.

**Comment by bluca** [comment]
The parameters passed to the polkit CheckAuthorization D-Bus call are not used to do authentication, they are used for identification. Authentication is done by polkit itself, which using the policy language plus optional external tools can then make a decision and answer back yes/no, looking at the process again. The triplet passed to polkit here is used by polkit to identify the process at that point - that's why it takes 3 parameters, if you check the link I've provided, it used to be only one (pid) but that was proven broken, then it used to take two (pid + uid) but that was proven broken too, and then it took three (pid + uid + start time), but in a turn of event that nobody saw coming, that was proven broken too - because there was really no reliable way to identify a process, before pid fds were added - that's the reason they were added in the first place.

Hence, if you want to do auth by uid, that's absolutely fine - the policy check will do that. But passing the uid to CheckAuthorization doesn't do that. It simply causes polkit to trust that the process is really the one being authenticated using that triplet, even if it really, really cannot know it actually is, because it's provenly broken, with CVEs and whatnot.

**Comment by poettering** [comment]
But that's entirely fine. If some code runs under UID x then it can insert itself into any other process of UID x anyway, hence it's *enough* to identify the UID, and the PID/startttime are just decoration more or less.

it's a problem of you can trick polkit into doing auth with different UID than the process runs under, but it's entirely fine if the uid can fake the pids of its own user.

So, let's not be ore catholic than the pope here. 

And if anything this should be turned off on the PK side anyway, not our client side.

**Comment by bluca** [comment]
It's not fine at all. The point of this call is to uniquely identify a 'subject' (in polkit terms), and doing it this way is not sufficient, so it's not only meaningless, but worse: it pretends there's a security policy in place, when there isn't, as identification is just broken and can't be relied upon. And you know perfectly well how D-Bus APIs work, one can't just change them and break compatibility, so it has to stay. It doesn't mean it's a good idea to use it since we know better.

I didn't go through the effort of weaving pid fd support through it just for the lolz. I've added this API upstream especially for your use case. I don't understand what's the problem with using it now?

**Comment by poettering** [comment]
> It's not fine at all. The point of this call is to uniquely identify a 'subject' (in polkit terms), and doing it this way is not sufficient, so it's not only meaningless, but worse: it pretends there's a security policy in place, when there isn't, as identification is just broken and can't be relied upon. And you know perfectly well how D-Bus APIs work, one can't just change them and break compatibility, so it has to stay. It doesn't mean it's a good idea to use it since we know better.
> 
> I didn't go through the effort of weaving pid fd support through it just for the lolz. I've added this API upstream especially for your use case. I don't understand what's the problem with using it now?

I don't understand a word. Are you really saying that auth-by-uid is not good enough for auth purposes?

if you auth by pidfd, then via ptrace/criu and stuff any other process can take possession of your process as long as it shares the same UID, hence the process binding is nice to route requests to the right agent, and it's nice for debugging, but it's not suitable as a security concept, while the UID definitely is.

**Comment by bluca** [comment]
> > It's not fine at all. The point of this call is to uniquely identify a 'subject' (in polkit terms), and doing it this way is not sufficient, so it's not only meaningless, but worse: it pretends there's a security policy in place, when there isn't, as identification is just broken and can't be relied upon. And you know perfectly well how D-Bus APIs work, one can't just change them and break compatibility, so it has to stay. It doesn't mean it's a good idea to use it since we know better.
> > I didn't go through the effort of weaving pid fd support through it just for the lolz. I've added this API upstream especially for your use case. I don't understand what's the problem with using it now?
> 
> I don't understand a word. Are you really saying that auth-by-uid is not good enough for auth purposes?

No, I am saying that calling CheckAuthorization(pid, uid, start_time) is _not_ auth-by-id. auth-by-id or by-anything-else is decided when the policy rules are evaluated, and it's entirely orthogonal to which parameters are passed to this call.

polkit works like this: it builds a representation of the "thing" that requires auth, and it calls a "subject". For all intents and purposes this is a process in our case. The subject needs to be _uniquely_ and _reliably_ identified - the way we do it normally is via D-Bus peer id, as the broker allows us to do that nicely and safely. Then, once the subject is set up with all its identifying parameters and attributes - which can and does happen ONCE and be reused MANY TIMES due to caching - the action id is used to evaluate rules, at which point many things can happen: one of this is, a rule that says "auth-by-uid", and if the uid of the subject matches what the rule allows, then all is good and the action is authorize for the subject. Rules can be arbitrarily complicated and ask for interactive user approval, or check by group name or id, or shell out to external helpers to get an answer, or (with the next version) match on a specific unit name, or a mix of all these things and more.

The crucial thing is, building the subject is not the same as authorizing it. As the code comment linked above describes, and as the many CVEs in this area prove, identifying a subject by pid[+uid[+start time]] is not safe against various attacks, and is deprecated (but it's a D-Bus interface so it will exist forever). Identifying by D-Bus peer id is fine because the broker keeps track of things. But that doesn't help when D-Bus is not used by the subject - hence why I added pid fd support for building a subject, as that _is_ safe against these attacks as guaranteed by the kernel as you know already.

Hence, if instead of calling CheckAuthorization(pid, uid, start_time), CheckAuthorization(pidfd) is called, it doesn't mean that you lose the ability to authorize by uid (you can even pass directly the uid as well, if you like and you already know it and want to 'pin it' at subject build time in case the rule doesn't check for NoNewPrivs, if it's specified it will be used, otherwise polkit will derive it from the process via pidfd when processing the call). Then, you install a rule that says "auth by this or that uid", and you get auth-by-uid, regardless of how the subject was put together, but without being open to all the various pid reuse attacks and races.

**Comment by poettering** [comment]
> CheckAuthorization(pidfd)

Sorry, but I hope noone ever does that, that's simply not safe. you cannot safely derive a UID from that, which makes this useless for authentication. We should *never* just pass a pidfd around, it must always be primarily a uid, and then some decoration around it useful for routing agent requests.

**Comment by poettering** [comment]
> case the rule doesn't check for NoNewPrivs

i am sorry what? rules check for NNP? how is that safe? you cannot atomically check NNP and other process creds, hence NNP should never be used for authentication, because otherwise if you manage to invoke some suid binary which happens to enable NNP soon, then polkit might assume all way OK given the pidfd you gave it, but actually it authenticates the wrong uid.

remote NNP checks are *not* ok, and a security vulnerability.

**Comment by bluca** [comment]
> > case the rule doesn't check for NoNewPrivs
> 
> i am sorry what? rules check for NNP?

Yes, as in the system unit property `NoNewPrivileges=`, if the subject is part of a unit and the option is set, it will be available to rules evaluation as a bool

**Comment by poettering** [comment]
> The crucial thing is, building the subject is not the same as authorizing it. As the code comment linked above describes, and as the many CVEs in this area prove, identifying a subject by pid[+uid[+start time]] is not safe against various attacks, and is deprecated (but it's a D-Bus interface so it will exist forever). Identifying by D-Bus peer id is fine because the broker keeps track of things. But that doesn't help when D-Bus is not used by the subject - hence why I added pid fd support for building a subject, as that _is_ safe against these attacks as guaranteed by the kernel as you know already.

I am sorry, but please provide an example what kind of attack that passing just a pidfd to a peer is going to block that passing uid+pid+starttime is vulnerable to.

**Comment by poettering** [comment]
> Apart from the fact that, as already said, you can pass the uid together with the pidfd just fine if you want to - why would that be unsafe, and why would it be any different from the D-Bus peer identification case? We have never, ever specified the uid in the existing CheckAuthorization calls, that currently only pass `system-bus-name` for the peer name. How is identifying by peer id different than pid fd here?

dbus stores the uid + selinux label + gids at the moment of connection, derived securely from SO_PEERCRED + SO_PEERSEC + SO_PEERGROUP. if you give polkit a bus name to authenticate, then that's just another way to saying please use the SO_PEERCRED/SO_PEERSEC/SO_PEERGROUP data from this client at a very well defined point in time, that definitely predates the moment of invocation of the call.

now, of this SO_PEERCRED data only the uid/gid really matter, as that cannot be faked. the PID is decoration, but still useful to route agent requests,

in a way authenticating a bus name via polkit just a shortcut for providing uid/gid/auxgids/selinux label manually. services could do that too, to the same effect. 

As these 4 fields are the only securely retrievable information about a client it's the only thing that security rules should make security decisions on. Everything else, i.e. session, seat, cgroup info (the former just being derived from the latter) cannot be retrieved securely, because the kernel provides no api for it. You could read it off a pid or pidfd, but that won't help you because it would be out-of-sync with the data that SO_PEERCRED/SO_PEERSEC/SO_PEERGROUP provide you with. Using this would be insecure, since cgroup migration exists and execve() exists and suid exists and "sudo" exists, which might open a new session/cgroup.

**Comment by poettering** [comment]
> > I am sorry, but please provide an example what kind of attack that passing just a pidfd to a peer is going to block that passing uid+pid+starttime is vulnerable to.
> 
> [CVE-2019-6133](https://github.com/advisories/GHSA-7j6c-jrh9-mvqm)

That does explicitly say that UID wasn't checked. Hence not helpful?

Sorry, but the discussion really boils down to: *uid* is what we need to pass along, the pidfd vs. pid starttime is orthogonal to that.

**Comment by bluca** [comment]
> > Apart from the fact that, as already said, you can pass the uid together with the pidfd just fine if you want to - why would that be unsafe, and why would it be any different from the D-Bus peer identification case? We have never, ever specified the uid in the existing CheckAuthorization calls, that currently only pass `system-bus-name` for the peer name. How is identifying by peer id different than pid fd here?
> 
> dbus stores the uid + selinux label + gids at the moment of connection, derived securely from SO_PEERCRED + SO_PEERSEC + SO_PEERGROUP. if you give polkit a bus name to authenticate, then that's just another way to saying please use the SO_PEERCRED/SO_PEERSEC/SO_PEERGROUP data from this client at a very well defined point in time, that definitely predates the moment of invocation of the call.

Ok, as mentioned you can still pass the uid if you have it, and it will be added to the subject attribute just the same

**Comment by poettering** [comment]
> I didn't go through the effort of weaving pid fd support through it just for the lolz. I've added this API upstream especially for your use case. I don't understand what's the problem with using it now?

pidfd is nice an all, but not available in any released version/RPM or so. It's also not making anything more secure. Specifying the uid otoh *is* making things more secure, and is enough.

I am happy to add a pidfd to the set of creds we pass over once we have it, but it's not that important to make things secure. An if not combined with a uid it's *less* secure. that uid+pid+starttime would be.

**Comment by bluca** [comment]
> > > I am sorry, but please provide an example what kind of attack that passing just a pidfd to a peer is going to block that passing uid+pid+starttime is vulnerable to.
> > 
> > 
> > [CVE-2019-6133](https://github.com/advisories/GHSA-7j6c-jrh9-mvqm)
> 
> That does explicitly say that UID wasn't checked. Hence not helpful?

It is helpful because it proves that you cannot uniquely identify a process even if you have the full tuple. Adding the uid is a workaround that reduces the scope of the attack, but doesn't fully mitigate it - it simply says, within the same session all goes and there's no guarantee that what the user prompt authenticates is really what it thinks it is authenticating, only that it comes from the same session. Of course that's much better than before and it's understandable that in a pre-pid-fd world that workaround was added, but it's not exactly uncommon to have vulnerabilities that grant local same-user execution privileges, eg; sandbox escape from a browser or flatpak or document reader or image viewer or whatever else. If you read the discussion, it was brainstormed to try and do better but there was simply no infrastructure available at the time. Now it's there.

> Sorry, but the discussion really boils down to: _uid_ is what we need to pass along, the pidfd vs. pid starttime is orthogonal to that.

Again, you can pass the uid just fine in either case. The point is that pid+starttime is _not_ equivalent to pidfd, and it's not good enough in 2023 for new features.

**Comment by bluca** [comment]
> > I didn't go through the effort of weaving pid fd support through it just for the lolz. I've added this API upstream especially for your use case. I don't understand what's the problem with using it now?
> 
> pidfd is nice an all, but not available in any released version/RPM or so.

This is not available in any released versions either, so it doesn't really matter. They will both be part of the next respective releases, and they will work together or not at all, as it's normal for new features across multiple components.

> It's also not making anything more secure. Specifying the uid otoh _is_ making things more secure, and is enough.
> I am happy to add a pidfd to the set of creds we pass over once we have it, but it's not that important to make things secure. An if not combined with a uid it's _less_ secure. that uid+pid+starttime would be.

No, it is most definitely not enough, as it still doesn't uniquely identify anything, it only tells you that it comes from the same session. That is not good enough. For example, one can define local rules that are more precise than 'same session', and then it all falls apart catastrophically.

**Comment by bluca** [comment]
> > > I didn't go through the effort of weaving pid fd support through it just for the lolz. I've added this API upstream especially for your use case. I don't understand what's the problem with using it now?
> > 
> > 
> > pidfd is nice an all, but not available in any released version/RPM or so.
> 
> This is not available in any released versions either, so it doesn't really matter. They will both be part of the next respective releases, and they will work together or not at all, as it's normal for new features across multiple components.

This relies on SO_PEERPIDFD which is also only available in the very latest kernel anyway, so it requires a new distro releases anyway

**Comment by poettering** [comment]
> It is helpful because it proves that you cannot uniquely identify a process even if you have the full tuple. Adding the uid is a workaround that reduces the scope of the attack, but doesn't fully mitigate it - it simply says, within the same session all goes and there's no guarantee that what the user prompt authenticates is really what it thinks it is authenticating, only that it comes from the same session. Of course that's much better than before and it's understandable that in a pre-pid-fd world that workaround was added, but it's not exactly uncommon to have vulnerabilities that grant local same-user execution privileges, eg; sandbox escape from a browser or flatpak or document reader or image viewer or whatever else. If you read the discussion, it was brainstormed to try and do better but there was simply no infrastructure available at the time. Now it's there.

So the only thing that pidfds really give you here is that you can auth the pidfd itself: i.e. when a pid comes once, and then comes back that that you can be sure it's definitely the same one. But that is a pretty weak functionality due to execve() of suid binaries, i.e. your pidfd might refer to something very different later. 

Other then that pidfds give you nothing valuable here, you cannot really derive cgroup membership and similar things securely from them, since they can change anytime asynchronously, and we make this accessible to unpriv clients via sudo/su, or systemd-run --scope. If you make this available anyway, you are just adding as security problem.

**Comment by poettering** [comment]
> > 
> 
> This PR relies on SO_PEERPIDFD which is also only available in the very latest kernel anyway, so it requires a new distro release anyway

my off-the-shelf fedora 39 has kernel 6.6. SO_PEERPIDFD was added in kernel 6.5.

**Comment by poettering** [comment]
> No, it is most definitely not enough, as it still doesn't uniquely identify anything, it only tells you that it comes from the same session. That is not good enough. For example, one can define local rules that are more precise than 'same session', and then it all falls apart catastrophically.

I am sorry, but *how* can you define something "more precise"? pidfds are not useful for that, they do not pin creds, they are a reference to a process which can mutate its credentials identity completely freely.

**Comment by poettering** [comment]
So I found this now: https://gitlab.freedesktop.org/polkit/polkit/-/blob/master/src/polkitbackend/polkitbackendjsauthority.cpp?ref_type=heads#L623

Which I guess is what you added there. While I think this information is useful to improve logging and things, it's not useful for binding authentication to, because it's acquired at a point the user has control over, so that the user can relatively easily game it. I think we can ignore it for security decisions, people have to do auth-by-uid or auth-by-gid right now on linux.

If linux had SO_PEERCGROUPID or so, things would be different, because from the cgroup we can securely derive session, and from the session we can securely derive the seat (since a sessoin can never migrate to a different seat). but the way this is implemented is insecure as fuck if used for auth decisions.

**Comment by bluca** [comment]
> > It is helpful because it proves that you cannot uniquely identify a process even if you have the full tuple. Adding the uid is a workaround that reduces the scope of the attack, but doesn't fully mitigate it - it simply says, within the same session all goes and there's no guarantee that what the user prompt authenticates is really what it thinks it is authenticating, only that it comes from the same session. Of course that's much better than before and it's understandable that in a pre-pid-fd world that workaround was added, but it's not exactly uncommon to have vulnerabilities that grant local same-user execution privileges, eg; sandbox escape from a browser or flatpak or document reader or image viewer or whatever else. If you read the discussion, it was brainstormed to try and do better but there was simply no infrastructure available at the time. Now it's there.
> 
> So the only thing that pidfds really give you here is that you can auth the pidfd itself: i.e. when a pid comes once, and then comes back that that you can be sure it's definitely the same one. But that is a pretty weak functionality due to execve() of suid binaries, i.e. your pidfd might refer to something very different later.
> 
> Other then that pidfds give you nothing valuable here, you cannot really derive cgroup membership and similar things securely from them, since they can change anytime asynchronously, and we make this accessible to unpriv clients via sudo/su, or `systemd-run --scope --user`. If you make this available anyway, you are just adding a security problem.

Again, what it gives that your current version with pid+starttime cannot give, is assurance that the process being checked is really the one intended to be checked and not a different one with a collided pid. Everything else you said applies just the same to pid+starttime or to pidfd, there's no difference.

**Comment by bluca** [comment]
> So I found this now: https://gitlab.freedesktop.org/polkit/polkit/-/blob/master/src/polkitbackend/polkitbackendjsauthority.cpp?ref_type=heads#L623
> 
> Which I guess is what you added there.

I did not add the logind session stuff if that's what you mean, it was already there. I added pidfd, NNP and system unit

> While I think this information is useful to improve logging and things, it's not useful for binding authentication to, because it's acquired at a point the user has control over, so that the user can relatively easily game it. I think we can ignore it for security decisions, people have to do auth-by-uid or auth-by-gid right now on linux.
> 
> If linux had SO_PEERCGROUPID or so, things would be different, because from the cgroup we can securely derive session, and from the session we can securely derive the seat (since a sessoin can never migrate to a different seat). but the way this is implemented is insecure as fuck if used for auth decisions.

If all one needed was uid/gid, polkit would be pretty much redundant, there's no need for it for just that. The point of polkit is to allow rich, custom and specific rules to be written, that can do things like require user authentication via GUI prompts for example. That's an entirely legitimate, existing and widespread use case, and it's not going to go away. And that's one more reason why correctly identifying subjects is important.

**Comment by poettering** [comment]
frankly, if you are using the seat/session/cgroup info in your "rich" rules to make decisions, then this is a security issue right now, and is not an excuse for anything.

Anyway, not sure what I should do with this now. You are building security decisions on unsafe stuff (pidfd-derived cgroups info), and then are upset if I avoid it, and instead provide polkit with safe stuff (uid). I really don't think your usecase makes sense, but if you want to block this because of that broken usecase, then fine, I'll prep a new PR that does the systemd-creds without any polkit hookup and we can close this PR here, and not do polkit for the forseeable future.

And frankly the seats/session/cgroup based logic in polkit should be killed, and in any reasonable world that stuff would get a CVE.

If people want to make session/seat based decision they should bind it back to the uid, i.e. rather than ask "does this process belong to seat X", but instead ask "does the user have a session on seat X". Because the latter is something one can safely answer with current APIs, the former certainly no.

**Comment by bluca** [comment]
I'm not using seat/session/cgroup, so it's not 'my' use case, as mentioned those are pre-existing. Those are just examples of things that are supported and need to work - and it's just an example, because polkit fully supports calling custom binaries to make decisions, so nobody can know what those are doing, so setting up the subject correctly is fundamental to ensure everything works as expected. Depending on your security model checking by uid might be the sufficient, insufficient or unnecessary, it really depends on who writes and ships the _rules_, not who writes the identification calls. It's entirely understandable and supported to do authentication between different uids too.
If you don't want to do the CheckAuth(pidfd, uid) change I can do it myself separately, that's not a problem.

File: src/shared/bus-polkit.c [hunk 1]
```diff
@@ -588,3 +601,238 @@ Hashmap *bus_verify_polkit_async_registry_free(Hashmap *registry) {
...
+        if (r < 0)
+                return r;
+
+        r = sd_bus_message_append(
+                        c,
+                        "(sa{sv})s",
+                        "unix-process", 3,
+                        "pid", "u", (uint32_t) pidref.pid,
+                        "start-time", "t", (uint32_t) start_time,
+                        "uid", "u", (uint32_t) uid,
```
Comment (bluca) [inline src/shared/bus-polkit.c]: `uid` is actually of type int32_t for this call (it's uint32 in different calls), so this cannot work and gets rejected https://gitlab.freedesktop.org/polkit/polkit/-/blob/123/src/polkit/polkitsubject.c?ref_type=tags#L433

**Comment by bluca** [comment]
> > > Apart from the fact that, as already said, you can pass the uid together with the pidfd just fine if you want to - why would that be unsafe, and why would it be any different from the D-Bus peer identification case? We have never, ever specified the uid in the existing CheckAuthorization calls, that currently only pass `system-bus-name` for the peer name. How is identifying by peer id different than pid fd here?
> > 
> > 
> > dbus stores the uid + selinux label + gids at the moment of connection, derived securely from SO_PEERCRED + SO_PEERSEC + SO_PEERGROUP. if you give polkit a bus name to authenticate, then that's just another way to saying please use the SO_PEERCRED/SO_PEERSEC/SO_PEERGROUP data from this client at a very well defined point in time, that definitely predates the moment of invocation of the call.
> 
> Ok, as mentioned you can still pass the uid if you have it, and it will be added to the subject attribute just the same

PR that enforces they get passed together or not at all: https://gitlab.freedesktop.org/polkit/polkit/-/merge_requests/205

File: src/shared/bus-polkit.c (hunk 1 above)
Comment (poettering) [reply src/shared/bus-polkit.c]: hm? this works perfectly here. the dialog pops up as it should, evertyhing works.

Comment (poettering) [reply src/shared/bus-polkit.c]: ah, i guess it doesn't get reject, just ignored. what a stupid broken interface...

Comment (bluca) [reply src/shared/bus-polkit.c]: yeah, it's really silly

**Comment by bluca** [comment]
> > > > Apart from the fact that, as already said, you can pass the uid together with the pidfd just fine if you want to - why would that be unsafe, and why would it be any different from the D-Bus peer identification case? We have never, ever specified the uid in the existing CheckAuthorization calls, that currently only pass `system-bus-name` for the peer name. How is identifying by peer id different than pid fd here?
> > > 
> > > 
> > > dbus stores the uid + selinux label + gids at the moment of connection, derived securely from SO_PEERCRED + SO_PEERSEC + SO_PEERGROUP. if you give polkit a bus name to authenticate, then that's just another way to saying please use the SO_PEERCRED/SO_PEERSEC/SO_PEERGROUP data from this client at a very well defined point in time, that definitely predates the moment of invocation of the call.
> > 
> > 
> > Ok, as mentioned you can still pass the uid if you have it, and it will be added to the subject attribute just the same
> 
> PR that enforces they get passed together or not at all: https://gitlab.freedesktop.org/polkit/polkit/-/merge_requests/205

That is now merged too

**Comment by poettering** [comment]
Dropping "replaced-by-other-pr", since this PR is actually quite different from #30578 and has various changes missing from that PR.

**Comment by bluca** [comment]
> Dropping "replaced-by-newer-pr", since this PR is actually quite different from #30578 and has various changes missing from that PR.

Rebased that PR now to pick up all changes - the difference is pidfd as before, the json_sensitive fixes and using auth_admin_keep instead of auth_admin (it is plausible that one might do bulk encrypt/decrypt of creds in batches, no point in re-auth again and again in a short interval)

**Comment by poettering** [comment]
I rebased this PR now on top of git main, which makes it work on real polkit versions.

**Comment by bluca** [comment]
> I rebased this PR now on top of git main, which makes it work on real polkit versions.

Polkit that speaks pidfd (v124) is now available in Fedora 40/Rawhide, Debian testing/unstable, Archlinux and Ubuntu 24.04. So when we release v256 with the varlink polkit integration, it will work everywhere immediately. Is this enough to close this now?

**Comment by DaanDeMeyer** [comment]
@bluca Don't forget OpenSUSE!

**Comment by bluca** [comment]
Tumbleweed is lagging a bit behind yes, but even systemd is at v254 still, so shouldn't be an issue

**Comment by poettering** [comment]
well, i don't run rawhide, and love to test my stuff inetractively before i test it. interactive auth is kinda what pk is about. I regularly include the remaining patch in my PRs to test things with. For example in #26826 i included in the "@poettering
[lennarts debug logic](https://github.com/systemd/systemd/pull/26826/commits/4f13880178c3c96503fbb162810acef15ebc3c74)" commit (which is immediately reverted again).

**Comment by bluca** [comment]
It's in Fedora 40 too, no? Not just rawhide (just looking at the website, I don't run Fedora)

**Comment by DaanDeMeyer** [comment]
@bluca Fedora 40 is rawhide at the moment. It's not released yet

**Comment by bluca** [comment]
Ah, got it - I thought it got numbered when it released. Can the package be pulled from 40 to 39 locally?

**Comment by keszybz** [comment]
> Ah, got it - I thought it got numbered when it released. Can the package be pulled from 40 to 39 locally?

No. It's built against newer deps, so it wouldn't be installable in an older release.

---

The status is that polkit-124 is available in a stable Fedora version.

So the title of this PR is now outdated, and the question is: do we want to provide compatibility with older polkit versions, e.g. Fedora 39.

FWIW, I think @bluca is/was being a bit overzealous here, but at this point, I don't see a strong reason to merge this patch. Things should "just work" at the point where systemd-256 is released, and if people want to use newest systemd on older distros, they can apply this patch locally.

Close?

**Comment by poettering** [comment]
Yeah, i guess we can close this now.

I still think bluca was wrong on this one, but i guess time made the issue obsolete.

**Pull request closed without merging**